### ISSUE TITLE
[java] Typeresolution fails on a simple primary when the source is loaded from a class literal

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#510]
Hi, I'm working on a temporary solution to resolve method usages until we change the AST.

While trying to test it, type resolution failed with an `IllegalStateException` on a very simple primary (`object.toString()`). Here's the full stack trace of the exception:
```
java.lang.IllegalStateException: None of the maximally specific methods are abstract.
net.sourceforge.pmd.lang.java.typeresolution.MethodType@16b3fc9e
net.sourceforge.pmd.lang.java.typeresolution.MethodType@e2d56bf

	at net.sourceforge.pmd.lang.java.typeresolution.MethodTypeResolution.selectAmongMaximallySpecific(MethodTypeResolution.java:289)
	at net.sourceforge.pmd.lang.java.typeresolution.MethodTypeResolution.selectMostSpecificMethod(MethodTypeResolution.java:259)
	at net.sourceforge.pmd.lang.java.typeresolution.MethodTypeResolution.getBestMethodReturnType(MethodTypeResolution.java:230)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:452)
	at net.sourceforge.pmd.lang.java.ast.ASTName.jjtAccept(ASTName.java:35)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:314)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:967)
	at net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix.jjtAccept(ASTPrimaryPrefix.java:42)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:810)
	at net.sourceforge.pmd.lang.java.ast.ASTPrimaryExpression.jjtAccept(ASTPrimaryExpression.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:230)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:670)
	at net.sourceforge.pmd.lang.java.ast.ASTExpression.jjtAccept(ASTExpression.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:174)
	at net.sourceforge.pmd.lang.java.ast.ASTVariableInitializer.jjtAccept(ASTVariableInitializer.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	...
```

The code samples that trigger the exception are [here](https://github.com/oowekyala/pmd/blob/62cedae0ca9204fc224f26bf9a154ecc46b19e59/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/UsageResolverTest.java#L32). I think it's notable that the first test throws the exception but the second doesn't, the only difference being that in the first case, the source is loaded using a class literal and in the second it's provided as a string. The methods used to parse and visit the source are taken from `ClassTypeResolverTest`.

**Comment by jsotuyod** [comment]
@WinterGrascph we should take a look at this whenever you can.

**Comment by jsotuyod** [comment]
@WinterGrascph bonus track: `MethodType` certainly needs a `toString()` for debug purposes.

**Comment by jsotuyod** [comment]
I just added a `toString` on 7cfc44abf8c8a8efc38332c029e3b25aa8cbda73, can probably have a better implementation, but will work for the time being.

**Comment by WinterGrascph** [comment]
The problem stems from method type resolution being incomplete. It does not yet detect if a method is overridden or shadowed by another, thus it adds the original method and the overridden method to the potentially applicable list of methods and since they have the same erasure, the piece of code which selects the maximally specific method looks for the abstract one, but since none of them are abstract it throws and exception (it's implementation specific which overload gets called according to the JLS if two non-abstract methods with the same erasure are maximally specific)

**Comment by jsotuyod** [comment]
@WinterGrascph we should probably not fail while we build the missing pieces, but simply leave the type as null. Otherwise, we may be hampering other developers working on separate areas of PMD

**Comment by WinterGrascph** [comment]
#513 is fix for now.

**Comment by oowekyala** [comment]
@jsotuyod @WinterGrascph Thanks for the quick reaction!

**Comment by oowekyala** [comment]
Okay I get a different exception now with the same tests:

```
java.lang.NullPointerException
	at net.sourceforge.pmd.lang.java.typeresolution.MethodTypeResolution.checkSubtypeability(MethodTypeResolution.java:52)
	at net.sourceforge.pmd.lang.java.typeresolution.MethodTypeResolution.selectMostSpecificMethod(MethodTypeResolution.java:257)
	at net.sourceforge.pmd.lang.java.typeresolution.MethodTypeResolution.getBestMethodReturnType(MethodTypeResolution.java:230)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:452)
	at net.sourceforge.pmd.lang.java.ast.ASTName.jjtAccept(ASTName.java:35)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:10)
	at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit(JavaParserVisitorAdapter.java:314)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:967)
	at net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix.jjtAccept(ASTPrimaryPrefix.java:42)
	at net.sourceforge.pmd.lang.java.typeresolution.ClassTypeResolver.visit(ClassTypeResolver.java:810)
	at net.sourceforge.pmd.lang.java.ast.ASTPrimaryExpression.jjtAccept(ASTPrimaryExpression.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
```

**Comment by adangel** [comment]
This might be "fixed" with #653: https://github.com/pmd/pmd/pull/653/commits/1582e40055edf29da36629b251dc293378c2efd1

At least, the NPEs are prevented...

**Comment by adangel** [comment]
A minimal overriding detection is in place: https://github.com/pmd/pmd/pull/653/files#diff-8d9049b8b2d522b7ab5128df5ef94423R466

This means, that the correct method should be selected in the first phase. In theory, we could throw again the exception, but I didn't try it out.
